tower-http/src/request_id.rs

```diff
@@ -0,0 +1,624 @@
+        let this = self.project();
+        let mut response = futures_core::ready!(this.inner.poll(cx))?;
+
+        if let Some(current_id) = response.headers().get(&*this.header_name) {
+            if response.extensions().get::<RequestId>().is_none() {
+                let current_id = current_id.clone();
+                response.extensions_mut().insert(RequestId::new(current_id));
+            }
+        } else if let Some(request_id) = this.request_id.take() {
+            response
+                .headers_mut()
+                .insert(this.header_name.clone(), request_id.0.clone());
+            response.extensions_mut().insert(request_id);
+        }
+
+        Poll::Ready(Ok(response))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::ServiceBuilderExt as _;
+    use hyper::{Body, Response};
+    use std::{
+        convert::Infallible,
+        sync::{
+            atomic::{AtomicU64, Ordering},
+            Arc,
+        },
+    };
+    use tower::{ServiceBuilder, ServiceExt};
+
+    #[allow(unused_imports)]
+    use super::*;
+
+    #[tokio::test]
+    async fn basic() {
+        let svc = ServiceBuilder::new()
+            .set_x_request_id(Counter::default())
+            .propagate_x_request_id()
+            .service_fn(handler);
+
+        // header on response
+        let req = Request::builder().body(Body::empty()).unwrap();
+        let res = svc.clone().oneshot(req).await.unwrap();
+        assert_eq!(res.headers()["x-request-id"], "0");
+
+        let req = Request::builder().body(Body::empty()).unwrap();
+        let res = svc.clone().oneshot(req).await.unwrap();
+        assert_eq!(res.headers()["x-request-id"], "1");
+
+        // doesn't override if header is already there
+        let req = Request::builder()
+            .header("x-request-id", "foo")
+            .body(Body::empty())
+            .unwrap();
+        let res = svc.clone().oneshot(req).await.unwrap();
+        assert_eq!(res.headers()["x-request-id"], "foo");
+
+        // extension propagated
+        let req = Request::builder().body(Body::empty()).unwrap();
+        let res = svc.clone().oneshot(req).await.unwrap();
+        assert_eq!(res.extensions().get::<RequestId>().unwrap().0, "2");
+    }
+
+    #[tokio::test]
+    async fn other_middleware_setting_request_id() {
+        let svc = ServiceBuilder::new()
+            .override_request_header(
+                HeaderName::from_static("x-request-id"),
+                HeaderValue::from_str("foo").unwrap(),
+            )
+            .set_x_request_id(Counter::default())
+            .map_request(|request: Request<_>| {
+                // `set_x_request_id` should set the extension if its missing
+                assert_eq!(request.extensions().get::<RequestId>().unwrap().0, "foo");
+                request
+            })
+            .propagate_x_request_id()
+            .service_fn(handler);
+
+        let req = Request::builder()
+            .header("x-request-id", "foo")
```

Comment:
Just to clarify, its expected to produce the same result even if this would have `bar` as header value here right?

Reply:
Yep. I've changed the test in 69692cec3744f to make that more clear.